example/counterBasedRng/src/counterBasedRng.cpp

```diff
@@ -0,0 +1,258 @@
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+ * IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <alpaka/alpaka.hpp>
+#include <alpaka/example/ExampleDefaultAcc.hpp>
+#include <alpaka/rand/RandPhiloxStateless.hpp>
+
+#include <chrono>
+#include <iostream>
+#include <random>
+#include <typeinfo>
+
+//! A kernel that fills an array with pseudo-random numbers.
+class CounterBasedRngKernel
+{
+public:
+    template<class TAcc>
+    using Vec = alpaka::Vec<alpaka::Dim<TAcc>, alpaka::Idx<TAcc>>;
+    template<class TAcc>
+    using Gen = typename alpaka::rand::PhiloxStateless4x32x10Vector<TAcc>;
+    template<class TAcc>
+    using Key = typename Gen<TAcc>::Key;
+    template<class TAcc>
+    using Counter = typename Gen<TAcc>::Counter;
+
+private:
+    template<unsigned int I>
+    struct ElemLoop
+    {
+        ALPAKA_NO_HOST_ACC_WARNING
+        template<typename TAcc, typename TElem>
+        static ALPAKA_FN_ACC auto elemLoop(
+            TAcc const& acc,
+            alpaka::experimental::
+                BufferAccessor<TAcc, TElem, alpaka::Dim<TAcc>::value, alpaka::experimental::WriteAccess> dst,
+            Key<TAcc> const& key,
+            Vec<TAcc> const& threadElemExtent,
+            Vec<TAcc>& threadFirstElemIdx) -> void
+        {
+            auto const threadLastElemIdx = threadFirstElemIdx[I] + threadElemExtent[I];
+            auto const threadLastElemIdxClipped
+                = (dst.extents[I] > threadLastElemIdx) ? threadLastElemIdx : dst.extents[I];
+
+            constexpr auto Dim = alpaka::Dim<TAcc>::value;
+
+            const auto firstElem = threadFirstElemIdx[I];
+            if constexpr(I < Dim - 1)
+            {
+                for(; threadFirstElemIdx[I] < threadLastElemIdxClipped; ++threadFirstElemIdx[I])
+                {
+                    ElemLoop<I + 1>::elemLoop(acc, dst, key, threadElemExtent, threadFirstElemIdx);
+                }
+            }
+            else
+            {
+                Counter<TAcc> c = {0, 0, 0, 0};
+                for(unsigned int i = 0; i < Dim; ++i)
+                    c[i] = threadFirstElemIdx[i];
+
+                for(; threadFirstElemIdx[Dim - 1] < threadLastElemIdxClipped; ++threadFirstElemIdx[Dim - 1])
+                {
+                    c[Dim - 1] = threadFirstElemIdx[Dim - 1];
+                    const auto random = Gen<TAcc>::generate(c, key);
+                    // to make use of the whole random vector we would need to ensure numElement[0] % 4 == 0
+                    dst[threadFirstElemIdx] = TElem(random[0]);
+                }
+            }
+            threadFirstElemIdx[I] = firstElem;
+        }
+    };
+
+public:
```

Comment:
Is this really code a user should write? If yes, then we should try to move some parts into alpaka and provide building blocks. If no, we should try to provide an easier example.

Reply:
Yes and No: This is code the user should write only if they want to fill an array with random number, but this is not a common use case. Usually, the user will generate the random number and use them in the same place. Writing it into an array is only a stand-in for mapping random numbers to extrinsic properties of your problem, here this is coordinates.

I do not see part of the RNG-related logic, that could be turned into generic building blocks here.

The only that could be improved would be to have an `operator=` to replace lines 70-72.